test/addons-napi/test_promise/test_promise.c

```diff
@@ -0,0 +1,65 @@
+#include <node_api.h>
+#include "../common.h"
+
+napi_ref deferred_ref = NULL;
+
+napi_value createPromise(napi_env env, napi_callback_info info) {
+  napi_value promise, deferred, in_one_call;
+  bool one_call;
+  size_t argc = 1;
+
+  NAPI_CALL(env, napi_get_cb_info(env, info, &argc, &in_one_call, NULL, NULL));
+  NAPI_CALL(env, napi_get_value_bool(env, in_one_call, &one_call));
+
+  if (deferred_ref != NULL) {
+    NAPI_CALL(env, napi_delete_reference(env, deferred_ref));
+  }
+
+  if (one_call) {
+    NAPI_CALL(env, napi_create_promise(env, &deferred, &promise));
+  } else {
+    NAPI_CALL(env, napi_create_promise(env, &deferred, NULL));
+    NAPI_CALL(env, napi_get_promise_from_deferred(env, deferred, &promise));
+  }
+  NAPI_CALL(env, napi_create_reference(env, deferred, 1, &deferred_ref));
+
+  return promise;
+}
```

Comment:
Of course, if we were to use the original interface of
```C
napi_create_promise(env, &promise, &resolve, &reject);
```
where `resolve` and `reject` are to be called with `napi_call_function()` then this problem would go away, but, like you [said](https://github.com/nodejs/node/pull/13717#issuecomment-309038821), this interface is preferable to that one.